nntrainer/utils/profiler.cpp

```diff
@@ -0,0 +1,220 @@
+    throw std::invalid_argument("event has never recorded");
+  }
+
+  return iter->second.back();
+}
+
+void GenericProfileListener::report(std::ostream &out) const {
+  const std::vector<unsigned int> column_size = {10, 23, 23, 23};
+  auto total_col_size =
+    std::accumulate(column_size.begin(), column_size.end(), 0);
+
+  if (warmups != 0) {
+    out << "warm up: " << warmups << '\n';
+  }
+
+  /// creating header
+  // clang-format off
+  out << std::setw(column_size[0]) << "key"
+      << std::setw(column_size[1]) << "avg"
+      << std::setw(column_size[2]) << "min"
+      << std::setw(column_size[3]) << "max" << '\n';
+  // clang-format on
+
+  // seperator
+  out << std::string(total_col_size, '=') << '\n';
+
+  /// calculate metrics while skipping warmups
+  for (auto &entry : time_taken) {
+    std::chrono::milliseconds avg{0};
+    std::chrono::milliseconds min{std::chrono::milliseconds::max()};
+    std::chrono::milliseconds max{std::chrono::milliseconds::min()};
+
+    unsigned int len = entry.second.size();
+
+    if (warmups >= len) {
+      out << std::left << std::setw(total_col_size) << event_to_str(entry.first)
+          << "less data then warmup\n";
+      continue;
+    }
+
+    for (unsigned int i = warmups; i < len; i++) {
+      auto &item = entry.second[i];
+      avg += item;
+      min = std::min(min, item);
+      max = std::max(max, item);
+    }
+    avg /= len;
+
+    // clang-format off
+    out << std::setw(column_size[0]) << event_to_str(entry.first)
+        << std::setw(column_size[1]) << avg.count()
+        << std::setw(column_size[2]) << min.count()
+        << std::setw(column_size[3]) << max.count() << '\n';
+    // clang-format on
+  }
+}
+
+Profiler &Profiler::Global() {
+  static Profiler instance;
+  return instance;
+}
+
+std::string event_to_str(const int event) {
+  switch (event) {
+  case EVENT::NN_FORWARD:
+    return "nn_forward";
+  case EVENT::TEMP:
+    return "temp";
+  }
+
+  std::stringstream ss;
+  ss << "undef(" << event << ')';
+  return ss.str();
+}
+
+void Profiler::start(const int &event) {
+  /// @todo: consider race condition
+  auto iter = start_time.find(event);
+
+  if (iter != start_time.end()) {
+    throw std::invalid_argument("profiler has already started");
+  }
+
+  start_time[event] = std::chrono::steady_clock::now();
+}
+
+void Profiler::end(const int &event) {
+  /// @todo: consider race condition
+  auto end = std::chrono::steady_clock::now();
+  auto iter = start_time.find(event);
+
+  if (iter == start_time.end()) {
+    throw std::invalid_argument("profiler hasn't started with the event");
+  }
+
+  auto duration =
+    std::chrono::duration_cast<std::chrono::milliseconds>(end - iter->second);
+  notify(event, duration);
+
+  start_time.erase(iter);
+}
+
+void Profiler::notify(const int &event,
+                      const std::chrono::milliseconds &value) {
+  for (auto &listener : all_event_listeners) {
```

Comment:
this must also be locked as this list can be modified while it notifies.

Reply:
However, this adds a big concern that notifies is called with every iteration (inside end) and that has a locking of its own, and I wonder, how much overhead it adds to the profiler/changes the values from the actual results.

Reply:
I'll add a lock for now.
Idea for the improvement:
0. Delay `onNotify()` call 
1. a `freeze`, `thaw` mechanism to control subscribe, (blocking unsubscribe) to ensure thread-safe behavior without locking in this function.
2. Use `shared_mutex` to relieve the problem (c++17 or boost) 

I'll leave this as an issue for now.. I'll keep in mind this can impact the performance.
If we have a performance issue, let's deal with it within this month.

Reply:
good.
shared_mutex is good. I am not sure if freeze/thaw can work, but can be checked later.